R/insight_request.R

```diff
@@ -0,0 +1,170 @@
+permissions and limitations under the License.
+
+This product includes software developed at data.world, Inc.
+https://data.world"
+
+#' Create insight object for new insights
+#' @param title Insight title
+#' @param description Optional insight description
+#' @param image_url URL of image representing the insight
+#' @param embed_url URL of content to embed
+#' @param markdown_body Markdown text containing the insight
+#' @param source_link Permalink to source code or platform this
+#' insight was generated with
+#' @param data_source_links List containing one or more permalinks
+#' to the data sources used to generate this insight
+#' @return Request object of type \code{insight_create_request}.
+#' @seealso \code{\link{create_insight}}
+#' @examples
+#' request <- dwapi::insight_create_request(title='A title',
+#'   description = 'A description',
+#'   image_url = 'https://site.org/image.png',
+#'   source_link = 'https://site.org/data')
+#' @export
+insight_create_request <-
+  function(title, description = NULL,
+           image_url = NULL, embed_url = NULL, markdown_body = NULL,
+           source_link = NULL, data_source_links = NULL) {
+
+    if (is.blank(title)) {
+      stop("title can't be null or empty")
+    }
+
+    if (all(is.blank(image_url), is.blank(embed_url),
+            is.blank(markdown_body))) {
+      stop("Must provide one of image_url, embed_url, or markdown_body")
+    }
+
+    if (sum(!is.blank(image_url), !is.blank(embed_url),
+            !is.blank(markdown_body)) > 1) {
+      stop("Must provide only one of image_url, embed_url, or markdown_body")
+    }
+
+    base_insight_request(title, description,
+                         image_url, embed_url,
+                         markdown_body,
+                         source_link,
+                         data_source_links,
+                         "insight_create_request")
+
+  }
+
+#' Create insight object for insights that replace existing insights
+#' @param title Insight title
+#' @param description Optional insight description
+#' @param image_url URL of image representing the insight
+#' @param embed_url URL of content to embed
+#' @param markdown_body Markdown text containing the insight
+#' @param source_link Permalink to source code or platform this
+#' insight was generated with
+#' @param data_source_links List containing one or more permalinks
+#' to the data sources used to generate this insight
+#' @return Request object of type \code{insight_create_request}.
+#' @seealso \code{\link{create_insight}}
+#' @examples
+#' request <- dwapi::insight_replace_request(title='A title',
+#'   description = 'A description',
+#'   image_url = 'https://site.org/image.png',
+#'   source_link = 'https://site.org/data')
+#' @export
+insight_replace_request <-
+  function(title, description = NULL,
+           image_url = NULL, embed_url = NULL, markdown_body = NULL,
+           source_link = NULL, data_source_links = NULL) {
+
+    if (is.blank(title)) {
```

Comment:
The reason this isn't in the base request is because the base request is also used for updates, where null values of the parameters are perfectly valid (and in fact indicate that a particular parameter should not be updated).  There is a minor code smell here, where the validation logic is repeated.  I can take care of that by pulling that logic out into a separate function, and just call it from the create and replace functions.